File: jde-mcp-ube-analyzer/src/tools/query.ts
import type { McpServer } from "@modelcontextprotocol/sdk/server/mcp.js";
import {
  JdeQueryTableSchema,
  type JdeQueryTableInput,
} from "../schemas/tools.js";
import { queryTable } from "../services/ais-client.js";
import { resolveColumns } from "../services/dictionary.js";
import { CHARACTER_LIMIT } from "../constants.js";

export function registerQueryTool(server: McpServer): void {
  server.registerTool(
    "jde_query_table",
    {
      title: "Query JDE Table",
      description: `Execute a read-only query against any JDE EnterpriseOne table via the AIS Data Service.

IMPORTANT: Before calling this tool, use jde_dictionary_search or jde_dictionary_table to discover valid table names and column aliases. JDE uses cryptic aliases (e.g. DOCO = Order Number, AN8 = Address Number) — guessing will fail.

This is a generic, flexible query tool. For common lookups, prefer the curated tools:
  - jde_sales_order_inquiry  → Sales order detail/header
  - jde_customer_lookup      → Address book / customer search
  - jde_item_availability    → Item branch on-hand

Args:
  - tableName (string): JDE table name (e.g. "F4211")
  - columns (string[]): JDE column aliases to return (e.g. ["DOCO","AN8","LITM","UORG"])
  - filters (array, optional): Filter conditions, each with column, operator, value
  - maxRows (number): Max rows to return (default 50, max 500)

Filter operators: EQUAL, NOT_EQUAL, LESS, LESS_EQUAL, GREATER, GREATER_EQUAL, BETWEEN, LIST, STR_CONTAIN, STR_START_WITH, STR_END_WITH, STR_BLANK, STR_NOT_BLANK

Returns:
  JSON with rows array and metadata (totalRecords, returnedRecords, hasMore).

Example:
  tableName: "F4211"
  columns: ["DOCO","LNID","AN8","LITM","UORG","UPRC","NXTR"]
  filters: [{ column: "AN8", operator: "EQUAL", value: "4242" }]`,
      inputSchema: JdeQueryTableSchema,
      annotations: {
        readOnlyHint: true,
        destructiveHint: false,
        idempotentHint: true,
        openWorldHint: true,
      },
    },
    async (params: JdeQueryTableInput) => {
      try {
        // Validate columns against dictionary (best-effort — warns but doesn't block)
        const { valid, invalid } = await resolveColumns(
          params.tableName,
          params.columns
        );

        let warning = "";
        if (invalid.length > 0) {
          warning =
            `⚠️ Unrecognized columns for ${params.tableName}: ${invalid.join(", ")}. ` +
            `They may still work if they exist in JDE but are not in the curated dictionary.\n\n`;
        }

        const response = await queryTable({
          tableName: params.tableName,
          columns: params.columns,
          filters: params.filters?.map((f) => ({
            column: f.column,
            operator: f.operator,
            value: f.value,
          })),
          maxRows: params.maxRows,
        });

        const gridData = response.fs_DATABROWSE?.data?.gridData;
        if (!gridData) {
          return {
            content: [{
              type: "text" as const,
              text: warning + "No data returned from AIS. The table may be empty or filters too restrictive.",
            }],
          };
        }

        const output = {
          tableName: params.tableName,
          totalRecords: gridData.summary.records,
          returnedRecords: gridData.rowset.length,
          hasMore: gridData.summary.moreRecords,
          rows: gridData.rowset,
        };

        let text = warning + JSON.stringify(output, null, 2);

        // Truncate if too large for context window
        if (text.length > CHARACTER_LIMIT) {
          text = text.slice(0, CHARACTER_LIMIT) +
            `\n\n... [TRUNCATED — ${gridData.summary.records} total records. Narrow your filters or reduce maxRows.]`;
        }

        return {
          content: [{ type: "text" as const, text }],
        };
      } catch (err: unknown) {
        const msg = err instanceof Error ? err.message : String(err);
        return {
          isError: true,
          content: [{
            type: "text" as const,
            text: `Error querying ${params.tableName}: ${msg}. ` +
              `Verify the table name and column aliases using jde_dictionary_search.`,
          }],
        };
      }
    }
  );
}

File: jde-mcp-python/tools/query.py
"""Layer 3 — Generic table query tool."""
import json
import logging
from typing import Optional
from mcp.server.fastmcp import FastMCP

from config import CHARACTER_LIMIT
from services.ais_client import query_table
from services.dictionary import resolve_columns

logger = logging.getLogger(__name__)


def register(mcp: FastMCP) -> None:

    @mcp.tool(
        name="jde_query_table",
        description=(
            "Execute a read-only query against any JDE EnterpriseOne table via AIS Data Service. "
            "IMPORTANT: Before calling, use jde_dictionary_search or jde_dictionary_table to discover valid table names "
            "and column aliases. JDE uses cryptic aliases (e.g. DOCO = Order Number, AN8 = Address Number) — guessing will fail. "
            "Args: table_name (string), columns (list of alias strings), "
            "filters (list of {column, operator, value}), max_rows (1-500, default 50). "
            "Operators: EQUAL, NOT_EQUAL, LESS, LESS_EQUAL, GREATER, GREATER_EQUAL, BETWEEN, LIST, "
            "STR_CONTAIN, STR_START_WITH, STR_END_WITH, STR_BLANK, STR_NOT_BLANK. "
            "Returns: JSON with rows array and metadata (totalRecords, returnedRecords, hasMore)."
        ),
        annotations={"readOnlyHint": True, "destructiveHint": False, "idempotentHint": True, "openWorldHint": True},
    )
    async def jde_query_table(
        table_name: str,
        columns: list[str],
        filters: Optional[list[dict]] = None,
        max_rows: int = 50,
    ) -> str:
        try:
            resolved = resolve_columns(table_name, columns)
            warning = ""
            if resolved["invalid"]:
                warning = (
                    f"Warning: Unrecognized columns for {table_name}: {', '.join(resolved['invalid'])}. "
                    "They may still work if they exist in JDE but are not in the curated dictionary.\n\n"
                )

            resp = await query_table(
                table_name=table_name,
                columns=columns,
                filters=[
                    {"column": f["column"], "operator": f["operator"], "value": f["value"]}
                    for f in (filters or [])
                ] or None,
                max_rows=max_rows,
            )

            grid = (resp.get("fs_DATABROWSE") or {}).get("data", {}).get("gridData")
            if not grid:
                return warning + "No data returned from AIS. The table may be empty or filters too restrictive."

            output = {
                "tableName": table_name,
                "totalRecords": grid.get("summary", {}).get("records"),
                "returnedRecords": len(grid.get("rowset", [])),
                "hasMore": grid.get("summary", {}).get("moreRecords"),
                "rows": grid.get("rowset", []),
            }
            text = warning + json.dumps(output, indent=2)
            if len(text) > CHARACTER_LIMIT:
                text = (
                    text[:CHARACTER_LIMIT]
                    + f"\n\n... [TRUNCATED — {grid.get('summary', {}).get('records')} total records. Narrow your filters or reduce max_rows.]"
                )
            return text
        except Exception as exc:
            raise RuntimeError(
                f"Error querying {table_name}: {exc}. Verify the table name and column aliases using jde_dictionary_search."
            ) from exc
